File: backend/services/stigmer-server/pkg/domain/workflowexecution/temporal/dispatch_test.go
package temporal

import (
	"testing"

	sessionv1 "github.com/stigmer/stigmer/apis/stubs/go/ai/stigmer/agentic/session/v1"
)

func TestResolveWorkflowTaskQueue(t *testing.T) {
	globalCfg := &Config{
		RunnerQueue:             "stigmer_runner",
		WorkflowActivityRouting: RoutingGlobal,
		DefaultExecutionTarget:  DefaultExecutionTargetLocal,
	}

	executionCfg := &Config{
		RunnerQueue:             "stigmer_runner",
		WorkflowActivityRouting: RoutingExecution,
		DefaultExecutionTarget:  DefaultExecutionTargetCloud,
	}

	t.Run("global routing always returns runner queue", func(t *testing.T) {
		result := ResolveWorkflowTaskQueue("wfx_test_123", sessionv1.ExecutionTarget_EXECUTION_TARGET_CLOUD, globalCfg)
		if result.TaskQueue != "stigmer_runner" {
			t.Errorf("expected stigmer_runner, got %q", result.TaskQueue)
		}
	})

	t.Run("global routing ignores execution_target", func(t *testing.T) {
		result := ResolveWorkflowTaskQueue("wfx_test_456", sessionv1.ExecutionTarget_EXECUTION_TARGET_UNSPECIFIED, globalCfg)
		if result.TaskQueue != "stigmer_runner" {
			t.Errorf("expected stigmer_runner, got %q", result.TaskQueue)
		}
	})

	t.Run("execution routing with CLOUD returns wfexec queue", func(t *testing.T) {
		result := ResolveWorkflowTaskQueue("wfx_cloud_789", sessionv1.ExecutionTarget_EXECUTION_TARGET_CLOUD, executionCfg)
		expected := "wfexec:wfx_cloud_789"
		if result.TaskQueue != expected {
			t.Errorf("expected %q, got %q", expected, result.TaskQueue)
		}
		if result.ExecutionTarget != sessionv1.ExecutionTarget_EXECUTION_TARGET_CLOUD {
			t.Errorf("expected CLOUD, got %v", result.ExecutionTarget)
		}
	})

	t.Run("execution routing with UNSPECIFIED resolves to CLOUD via config", func(t *testing.T) {
		result := ResolveWorkflowTaskQueue("wfx_unspec_aaa", sessionv1.ExecutionTarget_EXECUTION_TARGET_UNSPECIFIED, executionCfg)
		expected := "wfexec:wfx_unspec_aaa"
		if result.TaskQueue != expected {
			t.Errorf("expected %q, got %q", expected, result.TaskQueue)
		}
		if result.ExecutionTarget != sessionv1.ExecutionTarget_EXECUTION_TARGET_CLOUD {
			t.Errorf("expected CLOUD (from config default), got %v", result.ExecutionTarget)
		}
	})

	t.Run("execution routing with LOCAL returns global queue", func(t *testing.T) {
		result := ResolveWorkflowTaskQueue("wfx_local_bbb", sessionv1.ExecutionTarget_EXECUTION_TARGET_LOCAL, executionCfg)
		if result.TaskQueue != "stigmer_runner" {
			t.Errorf("expected stigmer_runner for LOCAL target, got %q", result.TaskQueue)
		}
		if result.ExecutionTarget != sessionv1.ExecutionTarget_EXECUTION_TARGET_LOCAL {
			t.Errorf("expected LOCAL, got %v", result.ExecutionTarget)
		}
	})

	t.Run("FormatWfExecTaskQueue produces correct format", func(t *testing.T) {
		queue := FormatWfExecTaskQueue("wfx_abc123")
		expected := "wfexec:wfx_abc123"
		if queue != expected {
			t.Errorf("expected %q, got %q", expected, queue)
		}
	})
}

File: backend/services/stigmer-server/src/temporal/workflowexecution/__tests__/dispatch.test.ts
/**
 * Pins resolveWorkflowTaskQueue against Go's dispatch_test.go: the two
 * routing modes, the wfexec:{id} format, and the UNSPECIFIED-target
 * resolution through the config's single rule.
 */
import { describe, expect, it } from "vitest";

import { ExecutionTarget } from "@stigmer/protos/ai/stigmer/agentic/session/v1/enum_pb";

import { createLogger } from "../../../boot/logger.js";
import {
  WORKFLOW_DEFAULT_EXECUTION_TARGET_CLOUD,
  WORKFLOW_DEFAULT_EXECUTION_TARGET_LOCAL,
  WORKFLOW_ROUTING_EXECUTION,
  WORKFLOW_ROUTING_GLOBAL,
  WorkflowExecutionTemporalConfig,
} from "../../../domain/workflowexecution/temporal/config.js";
import { resolveWorkflowTaskQueue } from "../dispatch.js";
import { formatWfExecTaskQueue } from "../names.js";

const silentLogger = createLogger({
  level: "error",
  pretty: false,
  write: () => {},
});

function config(
  routing: string,
  defaultTarget = WORKFLOW_DEFAULT_EXECUTION_TARGET_LOCAL,
): WorkflowExecutionTemporalConfig {
  return new WorkflowExecutionTemporalConfig(
    "workflow_execution_stigmer",
    "stigmer_runner",
    routing,
    defaultTarget,
  );
}

describe("resolveWorkflowTaskQueue", () => {
  it("routes to the global runner queue in global mode", () => {
    const result = resolveWorkflowTaskQueue(
      "wfe-1",
      ExecutionTarget.LOCAL,
      config(WORKFLOW_ROUTING_GLOBAL),
      silentLogger,
    );
    expect(result.taskQueue).toBe("stigmer_runner");
    expect(result.executionTarget).toBe(ExecutionTarget.LOCAL);
  });

  it("derives wfexec:{id} in execution mode regardless of target", () => {
    for (const target of [
      ExecutionTarget.LOCAL,
      ExecutionTarget.CLOUD,
      ExecutionTarget.UNSPECIFIED,
    ]) {
      const result = resolveWorkflowTaskQueue(
        "wfe-42",
        target,
        config(WORKFLOW_ROUTING_EXECUTION),
        silentLogger,
      );
      expect(result.taskQueue).toBe("wfexec:wfe-42");
    }
  });

  it("resolves UNSPECIFIED to the configured default target", () => {
    const local = resolveWorkflowTaskQueue(
      "wfe-1",
      ExecutionTarget.UNSPECIFIED,
      config(WORKFLOW_ROUTING_GLOBAL, WORKFLOW_DEFAULT_EXECUTION_TARGET_LOCAL),
      silentLogger,
    );
    expect(local.executionTarget).toBe(ExecutionTarget.LOCAL);

    const cloud = resolveWorkflowTaskQueue(
      "wfe-1",
      ExecutionTarget.UNSPECIFIED,
      config(WORKFLOW_ROUTING_GLOBAL, WORKFLOW_DEFAULT_EXECUTION_TARGET_CLOUD),
      silentLogger,
    );
    expect(cloud.executionTarget).toBe(ExecutionTarget.CLOUD);
  });

  it("passes explicit targets through unchanged", () => {
    const result = resolveWorkflowTaskQueue(
      "wfe-1",
      ExecutionTarget.CLOUD,
      config(WORKFLOW_ROUTING_GLOBAL, WORKFLOW_DEFAULT_EXECUTION_TARGET_LOCAL),
      silentLogger,
    );
    expect(result.executionTarget).toBe(ExecutionTarget.CLOUD);
  });

  it("formatWfExecTaskQueue matches Go's format", () => {
    expect(formatWfExecTaskQueue("abc")).toBe("wfexec:abc");
  });
});
